background/update.js

```diff
@@ -67,29 +80,53 @@ var updater = {
         updater.log(updater.SKIPPED + ` (${err}) #${style.id} ${style.name}`);
       });
 
-    function maybeFetchMd5(digest) {
-      if (!ignoreDigest && style.originalDigest && style.originalDigest !== digest) {
+    function checkIfEdited(digest) {
+      if (ignoreDigest) {
+        return;
+      }
+      if (style.originalDigest && style.originalDigest !== digest) {
         return Promise.reject(updater.EDITED);
       }
-      return download(style.md5Url);
     }
 
-    function maybeFetchCode(md5) {
-      if (!md5 || md5.length !== 32) {
-        return Promise.reject(updater.ERROR_MD5);
-      }
-      if (md5 === style.originalMd5 && style.originalDigest && !ignoreDigest) {
-        return Promise.reject(updater.SAME_MD5);
-      }
-      return download(style.updateUrl);
+    function maybeUpdateUSO() {
+      return download(style.md5Url).then(md5 => {
+        if (!md5 || md5.length !== 32) {
+          return Promise.reject(updater.ERROR_MD5);
+        }
+        if (md5 === style.originalMd5 && style.originalDigest && !ignoreDigest) {
+          return Promise.reject(updater.SAME_MD5);
+        }
+        return download(style.updateUrl)
+          .then(text => tryJSONparse(text));
+      });
+    }
+
+    function maybeUpdateUsercss() {
+      return download(style.updateUrl).then(text => {
+        const json = usercss.buildMeta(text);
+        const {usercssData: {version}} = style;
+        const {usercssData: {version: newVersion}} = json;
+        // re-install is invalid in a soft upgrade
+        if (semverCompare(version, newVersion) === 0 && !ignoreDigest) {
+          return Promise.reject(updater.SAME_VERSION);
+        }
+        // downgrade is always invalid
+        if (semverCompare(version, newVersion) > 0) {
+          return Promise.reject(updater.ERROR_VERSION);
+        }
+        return usercss.buildCode(json);
+      });
     }
 
-    function maybeSave(text) {
-      const json = tryJSONparse(text);
+    function maybeValidate(json) {
+      if (json.usercssData) {
+        // usercss is already validated while building
+        return json;
+      }
       if (!styleJSONseemsValid(json)) {
         return Promise.reject(updater.ERROR_JSON);
       }
-      json.id = style.id;
```

Comment:
Don't remove it. You're making duplicates of styles that only change the order of sections on the remote server (lines 94-96).

Reply:
:sigh: I don't like it but I guess I'll have to review every change in all files.

Reply:
I must have overlooked that `saveStyle` call. I guess we need to move the `styleSectionsEqual` check into `maybeSave` function. BTW, in which case the style would lack of `originalDigest` property?
https://github.com/eight04/stylus/blob/c2eadda70805699b4b6eb357cdbd7ec1ebffc277/background/update.js#L135-L137